# EDAP01 - Assignment Search: Solving a Maze Using Different Search Algorithms

In this assignment, you are supposed to code the following search algorithms to solve various given mazes:

    - Breadth-first search (BFS)
    - Depth-first search (DFS)
    - Uniform-cost search (UCS)
    - Greedy best-first search (GBFS)
    - A* search

Please follow the basic instructions below to complete this notebook. For each search algorithm, report the following findings:

    - The solution and its path cost
    - The total number of nodes expanded
    - The length of the frontier
    - The total memory size to store all nodes
    - The depth size 

Submit the notebook file with all outputs. Check that your submitted file is readable and contains all saved images. Document your code. Use comments in the code to explain how your implementation works and detail your design choices.


This assignment folder (**EDAP01-LabAssignmentSearch.zip**) contains the following content:
 
* `Assignment_Search.ipynb`: This is the current <a href="https://anaconda.org/anaconda/jupyter">Jupyter</a> notebook file that you are reading. It comes with instructions. You need to edit and submit this file.
* `images`: This directory contains the output images.
* `mazes`: This directory contains predefined mazes.
* `searchalgorithms`: This directory contains the Python scripts that you need to code in order to solve different mazes. Each search algorithm has its own Python file (e.g. bfs.py), which is instantiated from the base class **SearchAlgorithmBase**, as defined in base.py. Your algorithms are expected to use the properties and methods of **SearchAlgorithmBase**. You need to complete and submit these Python files in this folder.
* `src/maze.py`: This file contains the maze class with all the required methods to generate a maze. **You should not make any changes to this file.**   
* `main.py`: This is the main file for running your search algorithms on the given maze. **No need to make any changes to this file.** 
* `requirements.txt`: This file has the list of all required libs to be installed.

## Step 1 - Installing Python and  Jupyter Notebook

To work on assignments, you can use one of these environments listed below:

* Use the online service <a href="https://colab.research.google.com/">Google CoLab</a>. No additional installations are necessary. Please note that our animations may not work on Colab due to some incompatibility issues.
* Install <a href="https://code.visualstudio.com/download">Visual Studio Code</a>. It will prompt you to install all the necessary software when you open a notebook file.
* Install Python, <a href="https://jupyter.org/">JupyterLab</a> and all needed packages yourself using pip or conda. Alternatively, you can use <a href="https://anaconda.org/">Anaconda</a>   for easier installation.

## Step 2 - Learning Python and Jupyer Notebook
If you are not familiar with Python, then you should go through the provided Python tutorial to learn the basics of Python and its packages such as numpy. Some code examples that help with the assignments are already available in this assignment.

Many online tutorials, such as the <a href="https://www.dataquest.io/blog/jupyter-notebook-tutorial/">Jupyer notebook tutorial</a>, cover how to use Jupyter notebooks.  

## Step 3 - Installation of the Required Libs

Once you have set up the Python environment, you need to import [numpy] and [pygame](https://www.pygame.org/news). To install these additional libs, run the following command:

In [1]:
pip install -r requirements.txt

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.3.1 -> 26.0
[notice] To update, run: python.exe -m pip install --upgrade pip


## Step 4 - Sanity check

To make sure that you have the correct setup, lets run the following command below. It will run the search algoritm **example** defined in ./searchalgorithms/example.py on the fifth maze (i.e., maze_id=5). If everything works correctly, a new window will pop up showing the maze with the start and end points indicated in blue and red, respectively. Since this example code is incomplete, no path will be computed. Therefore, the code will output a message indicating that there are zero expanded nodes and cost. To close the new window, just press the **Esc** button. Note that the final version of the map with the found path will be automatically saved as an image in the ./images folder. You should then see the following image: ./images/maze_5_algorithm_example.png. 

In [1]:
!python main.py --maze_id=1 --search_algorithm_name="example"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
example is loaded
------------------------
Found Path with 0 cost
Expanded 0 nodes
Max Frontier size 0
Max node memory size 0
Max depth 0
------------------------


There are a total of five different mazes, each of which is shown below.

<table style="width:100%; border:none">
  <tr>
    <td style="text-align:center; border:none">
      <img src="./images/maze1.png" width="200" alt="maze_1"  align="center">
      <br>
      <b>1</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze2.png" width="200" alt="maze_2"  align="center">
      <br>
      <b>2</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze3.png" width="200" alt="maze_3"  align="center">
      <br>
      <b>3</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze4.png" width="200" alt="maze_4"  align="center">
      <br>
      <b>4</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/maze5.png" width="200" alt="maze_5"  align="center">
      <br>
      <b>5</b>
    </td>
  </tr>
</table>

Once you successfully complete the required search algorithms, you should see such animations below. These animations are for the fifth maze, with all visited nodes shown in blue, frontiers in red and the final shortest path in green. All obstacles such as walls are depicted in dark brown.
 
 <table style="width:100%; border:none">
  <tr> <img src="./images/legend.png" width="600"  align="left"> </tr>
  <tr>
    <td style="text-align:center; border:none">
      <img src="./images/5bfs.gif" width="200" alt="BFS"  align="center">
      <br>
      <b>BFS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5dfs.gif" width="200" alt="DFS"  align="center">
      <br>
      <b>DFS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5ucs.gif" width="200" alt="UCS"  align="center">
      <br>
      <b>UCS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5gbfs.gif" width="200" alt="GBS"  align="center">
      <br>
      <b>GBFS</b>
    </td>
    <td style="text-align:center; border:none">
      <img src="./images/5astar.gif" width="200" alt="A*"  align="center">
      <br>
      <b>A*</b>
    </td>
  </tr>
</table>

## Step 5 - Tasks

Please follow instructions below and complete each task. Submit your source codes (Python files under the /searchalgorithms folder) together with this notebook with your findings (e.g., tables fillout out below)!

### Task 1: Breadth-First Search 

Follow the pseudocode in the textbook/slides and implement the **Breadth-First Search (BFS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/bfs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your BFS algorithm in this specific file. 
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.


In [ ]:
!python main.py --maze_id=5 --search_algorithm_name="bfs"

In [4]:
# Run BFS on mazes 1 to 5
# wait for one to finish, then close it, the next will start automatically
for i in range(1, 6):
    !python main.py --maze_id={i} --search_algorithm_name="bfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
bfs is loaded
------------------------
Found Path with 62 cost
Expanded 346 nodes
Max Frontier size 14
Max node memory size 348
Max depth 63
------------------------
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
bfs is loaded
------------------------
Found Path with 104 cost
Expanded 372 nodes
Max Frontier size 9
Max node memory size 375
Max depth 105
------------------------
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
bfs is loaded
------------------------
Found Path with 36 cost
Expanded 143 nodes
Max Frontier size 9
Max node memory size 151
Max depth 37
------------------------
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https

__BFS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         | 62        | 346            | 63             | 348           | 14                |
| 2         | 104       | 372            | 105            | 375           | 9                 |
| 3         | 36        | 143            | 37             | 151           | 9                 |
| 4         | 58        | 695            | 59             | 701           | 28                | 
| 5         | 50        | 576            | 51             | 581           | 22                | 
| 6         |           |                |                |               |                   | 


### Task 2: Depth-First Search 

Follow the pseudocode in the textbook/slides and implement the **Depth-First Search (DFS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/dfs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your DFS algorithm in this specific file. 
* Note that you can also use the BFS code here. In this case, however, you need to switch from the **Queue** data structure to the **Stack** data structure. 
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [ ]:
!python main.py --maze_id=2 --search_algorithm_name="dfs"

In [5]:
# Run DFS on mazes 1 to 5
# wait for one to finish, then close it, the next will start automatically
for i in range(1, 6):
    !python main.py --maze_id={i} --search_algorithm_name="dfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
dfs is loaded
------------------------
Found Path with 100 cost
Expanded 308 nodes
Max Frontier size 9
Max node memory size 316
Max depth 100
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
dfs is loaded
------------------------
Found Path with 216 cost
Expanded 334 nodes
Max Frontier size 25
Max node memory size 356
Max depth 216
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
dfs is loaded
------------------------
Found Path with 42 cost
Expanded 450 nodes
Max Frontier size 49
Max node memory size 457
Max depth 146
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pyga

__DFS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         | 100       | 308            | 100            | 316           | 9                 |
| 2         | 216       | 334            | 216            | 356           | 25                |
| 3         | 42        | 450            | 146            | 457           | 49                |
| 4         | 216       | 505            | 246            | 616           | 139               | 
| 5         | 204       | 273            | 204            | 415           | 143               | 

### Task 3: Uniform-Cost Search 

Follow the pseudocode in the textbook/slides and implement the **Uniform-Cost Search (UCS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/ucs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your UCS algorithm in this specific file. 
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [12]:
!python main.py --maze_id=2 --search_algorithm_name="ucs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
ucs is loaded
------------------------
Found Path with 104 cost
Expanded 372 nodes
Max Frontier size 9
Max node memory size 375
Max depth 105
------------------------


In [6]:
# Run UCS on mazes 1 to 5
# wait for one to finish, then close it, the next will start automatically
for i in range(1, 6):
    !python main.py --maze_id={i} --search_algorithm_name="ucs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
ucs is loaded
------------------------
Found Path with 62 cost
Expanded 346 nodes
Max Frontier size 14
Max node memory size 348
Max depth 63
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
ucs is loaded
------------------------
Found Path with 104 cost
Expanded 372 nodes
Max Frontier size 9
Max node memory size 375
Max depth 105
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
ucs is loaded
------------------------
Found Path with 36 cost
Expanded 143 nodes
Max Frontier size 9
Max node memory size 151
Max depth 37
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame c

__UCS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         | 62        | 346            | 63             | 348           | 14                |
| 2         | 104       | 372            | 105            | 375           | 9                 |
| 3         | 36        | 143            | 37             | 151           | 9                 |
| 4         | 58        | 695            | 59             | 701           | 28                | 
| 5         | 50        | 576            | 51             | 581           | 22                |

### Task 4: Greedy Best-First Search 

Follow the pseudocode in the textbook/slides and implement the **Greedy Best-First Search (GBFS)** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/gbfs.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your GBFS algorithm in this specific file. Use the **Manhattan distance** as the heuristic function.
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [13]:
!python main.py --maze_id=2 --search_algorithm_name="gbfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
gbfs is loaded
------------------------
Found Path with 118 cost
Expanded 164 nodes
Max Frontier size 19
Max node memory size 182
Max depth 118
------------------------


In [7]:
# Run GBFS on mazes 1 to 5
# wait for one to finish, then close it, the next will start automatically
for i in range(1, 6):
    !python main.py --maze_id={i} --search_algorithm_name="gbfs"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
gbfs is loaded
------------------------
Found Path with 66 cost
Expanded 89 nodes
Max Frontier size 16
Max node memory size 103
Max depth 66
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
gbfs is loaded
------------------------
Found Path with 118 cost
Expanded 164 nodes
Max Frontier size 19
Max node memory size 182
Max depth 118
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
gbfs is loaded
------------------------
Found Path with 42 cost
Expanded 44 nodes
Max Frontier size 10
Max node memory size 53
Max depth 42
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame

__GBFS__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         | 66        | 89             | 66             | 103           | 16                |
| 2         | 118       | 164            | 118            | 182           | 19                |
| 3         | 42        | 44             | 42             | 53            | 10                |
| 4         | 58        | 151            | 58             | 195           | 45                | 
| 5         | 50        | 156            | 50             | 207           | 52                | 


### Task 5: A* Search 

Follow the pseudocode in the textbook/slides and implement the **A*** strategy.   

Read the following **important notes** carefully:
* See the file **./searchalgorithms/astar.py**  and the class instance there. Override inherited functions from SearchAlgorithmBase and implement other necessery methods for your A* algorithm in this specific file. Use the **Manhattan distance** as the heuristic function.
* Do not edit any other code in any other file.
* Run the following code for the first four mazes by simply increasing the maze_id number from 1 to 4 and complete the table below with the results obtained for each maze. To close the window, just press the **Esc** button.

In [14]:
!python main.py --maze_id=2 --search_algorithm_name="astar"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
astar is loaded
------------------------
Found Path with 104 cost
Expanded 279 nodes
Max Frontier size 9
Max node memory size 284
Max depth 104
------------------------


In [8]:
# Run A* on mazes 1 to 5
# wait for one to finish, then close it, the next will start automatically
for i in range(1, 6):
    !python main.py --maze_id={i} --search_algorithm_name="astar"

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze1.txt is loaded
astar is loaded
------------------------
Found Path with 62 cost
Expanded 181 nodes
Max Frontier size 17
Max node memory size 186
Max depth 62
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze2.txt is loaded
astar is loaded
------------------------
Found Path with 104 cost
Expanded 279 nodes
Max Frontier size 9
Max node memory size 284
Max depth 104
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze3.txt is loaded
astar is loaded
------------------------
Found Path with 36 cost
Expanded 86 nodes
Max Frontier size 10
Max node memory size 95
Max depth 36
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pyg

__A*__

| maze id | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| 1         | 62        | 181            |  62            | 186           | 17                |
| 2         | 104       | 279            |  104           | 284           | 9                 |
| 3         | 36        | 86             |  36            | 95            | 10                |
| 4         | 58        | 524            |  58            | 539           | 27                | 
| 5         | 50        | 413            |  50            | 450           | 44                | 

### Task 6: Comparison and discussion

Run all your implemented search algorithms on maze 5 (i.e. maze_id=5) and complete the following table with your findings. Discuss your reported scores in the table. For example, explain which method returns the lowest path cost and why. Likewise, explain which method is more memory friendly and why that is the case. If you have any more discussion points, please do share them. 

__Maze #5__

| algorithm | path cost | # of nodes expanded | max tree depth | max # of nodes in memory | max frontier size |
|-----------|-----------|----------------|----------------|---------------|-------------------|
| BFS       | 50        | 576            | 51             | 581           | 22                |
| DFS       | 204       | 273            | 204            | 415           | 143               |
| UCS       | 50        | 576            | 51             | 581           | 22                |
| GBFS      | 50        | 156            | 50             | 207           | 52                |
| A*        | 50        | 413            | 50             | 450           | 44                |

 

In [9]:
# running all algorithms on maze 5 for comparison
maze_id = 5
algorithms = ["bfs", "dfs", "ucs", "gbfs", "astar"]
for alg in algorithms:
    !python main.py --maze_id={maze_id} --search_algorithm_name={alg}

pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
bfs is loaded
------------------------
Found Path with 50 cost
Expanded 576 nodes
Max Frontier size 22
Max node memory size 581
Max depth 51
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
dfs is loaded
------------------------
Found Path with 204 cost
Expanded 273 nodes
Max Frontier size 143
Max node memory size 415
Max depth 204
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
./mazes/maze5.txt is loaded
ucs is loaded
------------------------
Found Path with 50 cost
Expanded 576 nodes
Max Frontier size 22
Max node memory size 581
Max depth 51
------------------------
Exit
pygame 2.6.1 (SDL 2.28.4, Python 3.13.1)
Hello from the pygam

Discuss the most important lessons you have learned from implementing these different search strategies. 

In [ ]:
# Your discussion comes here!

### Task 7: Advanced task only for PhD Students
This task is for **PhD Students only**. Undergraduate students do not need to complete this task.

Modify your A* search to add weights (see textbook and lecture slides) and explore how different weights influence the result. Remember to create a new Python file under ./searchalgorithms.